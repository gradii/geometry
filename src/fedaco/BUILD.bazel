load("@npm//scss-bundle:index.bzl", "scss_bundle")
load(
    "//src/fedaco:config.bzl",
    "FEDACO_TARGETS",
)
load("//tools:defaults.bzl", "ng_package", "ts_library")

package(default_visibility = ["//visibility:public"])

ts_library(
    name = "fedaco",
    srcs = glob(
        ["**/*.ts"],
        exclude = [
            "test/**/*.ts",
            "**/*.spec.ts",
        ],
    ),
    deps = [
       "@npm//date-fns",
       "//src/check-type:check-type",
    ],
    module_name = "@gradii/fedaco",
)

ng_package(
    name = "npm_package",
    srcs = [
        "package.json",
    ],
    data = [
    ],
    entry_point = ":index.ts",
    entry_point_name = "fedaco",
    nested_packages = [],
#    tags = ["release-package"],
    deps = FEDACO_TARGETS,
)
