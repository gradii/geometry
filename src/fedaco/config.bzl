FEDACO_ENTRYPOINTS = [
]

FEDACO_TARGETS = ["//src/fedaco"]
