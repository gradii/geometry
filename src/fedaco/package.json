{
  "name": "@gradii/fedaco",
  "version": "0.0.1-beta.4",
  "description": "fedaco orm",
  "repository": {
    "type": "git",
    "url": "https://github.com/gradii/fedaco.git"
  },
  "keywords": [
    "fedaco"
  ],
  "license": "MIT",
  "dependencies": {
    "@gradii/annotation": "^1.1.2",
    "@gradii/check-type": "^1.1.2",
    "ramda": "^0.27.1"
  },
  "peerDependencies": {
    "tslib": "0.0.0-TSLIB"
  },
  "sideEffects": false
}

