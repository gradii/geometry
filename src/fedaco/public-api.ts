/**
 * @licence
 * Copyright (c) 2018 LinBo Len <linbolen@gradii.com>
 *
 * Use of this source code is governed by an MIT-style license.
 * See LICENSE file in the project root for full license information.
 */

export * from './src/query-builder/query-builder';
export * from './src/query-builder/grammar.interface';