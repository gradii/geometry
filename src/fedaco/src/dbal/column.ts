/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license
 */


/*Object representation of a database column.*/
export class Column {
  /**/
  protected _type: Type;
  /**/
  protected _length: number | null = null;
  /**/
  protected _precision: number = 10;
  /**/
  protected _scale: number = 0;
  /**/
  protected _unsigned: boolean = false;
  /**/
  protected _fixed: boolean = false;
  /**/
  protected _notnull: boolean = true;
  /**/
  protected _default: string | null = null;
  /**/
  protected _autoincrement: boolean = false;
  /**/
  protected _platformOptions: any[] = [];
  /**/
  protected _columnDefinition: string | null = null;
  /**/
  protected _comment: string | null = null;
  /**/
  protected _customSchemaOptions: any[] = [];

  /*Creates a new Column.*/
  public constructor(columnName: string, type, options: any[] = []) {
    super();
    this._setName(columnName);
    this.setType(type);
    this.setOptions(options);
  }

  /**/
  public setOptions(options: any[]) {
    for (const [name, value] of Object.entries(options)) {
      const method = 'set' + name;

      if (typeof this[method] !== 'function') {
        trigger_error(`The "${name}" column option is not supported, setting it is deprecated and will cause an error in Doctrine DBAL 3.0`,
          E_USER_DEPRECATED);
        continue;
      }
      this[method](value);
    }
    return this;
  }

  /**/
  public setType(type) {
    this._type = type;
    return this;
  }

  /**/
  public setLength(length: number | null) {
    if (length !== null) {
      this._length =
        // cast type int
        length;
    } else {
      this._length = null;
    }
    return this;
  }

  /**/
  public setPrecision(precision: number) {
    if (!is_numeric(precision)) {
      const precision = 10;
    }
    this._precision =
      // cast type int
      precision;
    return this;
  }

  /**/
  public setScale(scale: number) {
    if (!is_numeric(scale)) {
      const scale = 0;
    }
    this._scale =
      // cast type int
      scale;
    return this;
  }

  /**/
  public setUnsigned(unsigned: boolean) {
    this._unsigned =
      // cast type bool
      unsigned;
    return this;
  }

  /**/
  public setFixed(fixed: boolean) {
    this._fixed =
      // cast type bool
      fixed;
    return this;
  }

  /**/
  public setNotnull(notnull: boolean) {
    this._notnull =
      // cast type bool
      notnull;
    return this;
  }

  /**/
  public setDefault(_default: any) {
    this._default = _default;
    return this;
  }

  /**/
  public setPlatformOptions(platformOptions: any[]) {
    this._platformOptions = platformOptions;
    return this;
  }

  /**/
  public setPlatformOption(name: string, value: any) {
    this._platformOptions[name] = value;
    return this;
  }

  /**/
  public setColumnDefinition(value: string) {
    this._columnDefinition = value;
    return this;
  }

  /**/
  public getType() {
    return this._type;
  }

  /**/
  public getLength() {
    return this._length;
  }

  /**/
  public getPrecision() {
    return this._precision;
  }

  /**/
  public getScale() {
    return this._scale;
  }

  /**/
  public getUnsigned() {
    return this._unsigned;
  }

  /**/
  public getFixed() {
    return this._fixed;
  }

  /**/
  public getNotnull() {
    return this._notnull;
  }

  /**/
  public getDefault() {
    return this._default;
  }

  /**/
  public getPlatformOptions() {
    return this._platformOptions;
  }

  /**/
  public hasPlatformOption(name: string) {
    return this._platformOptions[name] !== undefined;
  }

  /**/
  public getPlatformOption(name: string) {
    return this._platformOptions[name];
  }

  /**/
  public getColumnDefinition() {
    return this._columnDefinition;
  }

  /**/
  public getAutoincrement() {
    return this._autoincrement;
  }

  /**/
  public setAutoincrement(flag: boolean) {
    this._autoincrement = flag;
    return this;
  }

  /**/
  public setComment(comment: string | null) {
    this._comment = comment;
    return this;
  }

  /**/
  public getComment() {
    return this._comment;
  }

  /**/
  public setCustomSchemaOption(name: string, value: any) {
    this._customSchemaOptions[name] = value;
    return this;
  }

  /**/
  public hasCustomSchemaOption(name: string) {
    return this._customSchemaOptions[name] !== undefined;
  }

  /**/
  public getCustomSchemaOption(name: string) {
    return this._customSchemaOptions[name];
  }

  /**/
  public setCustomSchemaOptions(customSchemaOptions: any[]) {
    this._customSchemaOptions = customSchemaOptions;
    return this;
  }

  /**/
  public getCustomSchemaOptions() {
    return this._customSchemaOptions;
  }

  /**/
  public toArray() {
    return {
      ...{
        name            : this._name,
        type            : this._type,
        default         : this._default,
        notnull         : this._notnull,
        length          : this._length,
        precision       : this._precision,
        scale           : this._scale,
        fixed           : this._fixed,
        unsigned        : this._unsigned,
        autoincrement   : this._autoincrement,
        columnDefinition: this._columnDefinition,
        comment         : this._comment
      },
      ...this._platformOptions,
      ...this._customSchemaOptions
    };
  }
}
