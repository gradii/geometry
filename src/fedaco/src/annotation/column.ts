/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license
 */

import { makePropDecorator } from '@gradii/annotation';
import { Model } from '../fedaco/model';

export interface ColumnDecorator {

  (obj?: ColumnDefine): any;

  isTypeOf(obj: any): obj is ColumnDefine;

  metadataName: string;

  /**
   * See the `Pipe` decorator.
   */
  new(obj?: ColumnDefine): ColumnDefine;
}

export interface ColumnDefine {
  isPrimary?: boolean;

  columnName?: string;
  // hasGetter?: boolean;
  // hasSetter?: boolean;
  serializeAs?: string;
  /**
   * Invoked before serializing process happens
   */
  serialize?: (value: any, attribute: string, model: any) => any;
  /**
   * Invoked before create or update happens
   */
  prepare?: (value: any, attribute: string, model: any) => any;
  /**
   * Invoked when row is fetched from the database
   */
  consume?: (value: any, attribute: string, model: any) => any;

  meta?: any;

  isDate?: boolean;

  isDateCastable?: boolean;

  isEncryptedCastable?: boolean;

  isRelation?: boolean;

  isRelationUsing?: boolean;
}

export interface HasOneColumnDefine extends ColumnDefine {
  isRelation: boolean;

  // relationsDefinition: Relation;
}

const _additionalProcessing = (target: any, name: string, columnDefine: ColumnDefine) => {
  const descriptor = Object.getOwnPropertyDescriptor(target, name);

  // columnDefine.isPrimary   = columnDefine.isPrimary || false;
  // columnDefine.columnName  = columnDefine.columnName || snakeCase(name);
  // columnDefine.serializeAs = columnDefine.serializeAs != null ?
  //   columnDefine.serializeAs : snakeCase(name);

  const hasGetter = !!(descriptor && descriptor.get);
  const hasSetter = !!(descriptor && descriptor.set);

  if (!hasGetter || !hasSetter) {
    const propertyDescriptor: PropertyDescriptor = {
      enumerable  : false,
      configurable: true
    };
    if (!hasGetter) {
      propertyDescriptor.get = function () {
        return (this as Model).getAttribute(name);
      };
    }
    if (!hasSetter) {
      propertyDescriptor.set = function (value) {
        (this as Model).setAttribute(name, value);
      };
    }
    Object.defineProperty(target, name, propertyDescriptor);
  }
};

// tslint:disable-next-line:variable-name
export const Column: ColumnDecorator = makePropDecorator(
  'fedaco orm column',
  (p: ColumnDefine = {}) => ({...p}), undefined,
  (target: any, name: string, columnDefine: ColumnDefine) => {
    _additionalProcessing(target, name, columnDefine);
  });

// tslint:disable-next-line:variable-name
export const DateColumn: ColumnDecorator = makePropDecorator(
  'fedaco orm date column',
  (p: ColumnDefine = {}): ColumnDefine => ({...p, isDate: true}), undefined,
  (target: any, name: string, columnDefine: ColumnDefine) => {
    _additionalProcessing(target, name, columnDefine);
  });

// tslint:disable-next-line:variable-name
export const DateCastableColumn: ColumnDecorator = makePropDecorator(
  'fedaco orm date castable column',
  (p: ColumnDefine = {}): ColumnDefine => ({...p, isDate: false, isDateCastable: true}), undefined,
  (target: any, name: string, columnDefine: ColumnDefine) => {
    _additionalProcessing(target, name, columnDefine);
  });

// tslint:disable-next-line:variable-name
export const EncryptedCastableColumn: ColumnDecorator = makePropDecorator(
  'fedaco orm encrypted castable column',
  (p: ColumnDefine = {}): ColumnDefine => ({...p, isEncryptedCastable: true}), undefined,
  (target: any, name: string, columnDefine: ColumnDefine) => {
    _additionalProcessing(target, name, columnDefine);
  });
