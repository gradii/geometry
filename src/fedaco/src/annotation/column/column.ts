/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license
 */

import { makePropDecorator } from '@gradii/annotation';
import { _additionalProcessingGetterSetter } from '../additional-processing';
import { FedacoDecorator } from '../annotation.interface';
import { ColumnAnnotation, FedacoColumn } from '../column';

export const Column: FedacoDecorator<ColumnAnnotation> = makePropDecorator(
  'Fedaco:Column',
  (p: ColumnAnnotation = {}): ColumnAnnotation => ({...p}),
  FedacoColumn,
  (target: any, name: string, columnDefine: ColumnAnnotation) => {
    _additionalProcessingGetterSetter(target, name, columnDefine);
  });
