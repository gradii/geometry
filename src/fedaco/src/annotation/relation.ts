/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license
 */

import { makePropDecorator, TypeDecorator } from '@gradii/annotation';
import { Model } from '../fedaco/model';
import { ColumnDecorator, ColumnDefine } from './column';

export interface RelationDecorator {

  (obj?: RelationAnnotation): TypeDecorator;

  /**
   * See the `Pipe` decorator.
   */
  new(obj?: RelationAnnotation): RelationAnnotation;
}

export interface RelationAnnotation {
  name?: string;
}

const _additionalProcessing = (target: any, name: string, columnDefine: ColumnDefine) => {
  const descriptor = Object.getOwnPropertyDescriptor(target, name);

  // columnDefine.isPrimary   = columnDefine.isPrimary || false;
  // columnDefine.columnName  = columnDefine.columnName || snakeCase(name);
  // columnDefine.serializeAs = columnDefine.serializeAs != null ?
  //   columnDefine.serializeAs : snakeCase(name);

  const hasGetter = !!(descriptor && descriptor.get);
  const hasSetter = !!(descriptor && descriptor.set);

  if (!hasGetter || !hasSetter) {
    const propertyDescriptor: PropertyDescriptor = {
      enumerable  : false,
      configurable: true
    };
    if (!hasGetter) {
      propertyDescriptor.get = function () {
        return (this as Model).getAttribute(name);
      };
    }
    if (!hasSetter) {
      propertyDescriptor.set = function (value) {
        (this as Model).setAttribute(name, value);
      };
    }
    Object.defineProperty(target, name, propertyDescriptor);
  }
};


// ====
// @RelationColumn({
//  columnName: 'articles',
//  onQuery: ()=>{
//  }
// })
// @RelationUsingColumn({
//  columnName: 'articles',
//  callback: (model)=>{
//    return new EloquentResolverRelationStub(builder, $model);
//  }
// })
//
// tslint:disable-next-line:variable-name
export const RelationColumn: ColumnDecorator = makePropDecorator(
  'fedaco orm relation column',
  (p: ColumnDefine = {}): ColumnDefine => ({...p, isRelation: true}), undefined,
  (target: any, name: string, columnDefine: ColumnDefine) => {
    _additionalProcessing(target, name, columnDefine);
  });

export const RelationUsingColumn: ColumnDecorator = makePropDecorator(
  'fedaco orm relation using column',
  (p: ColumnDefine = {}): ColumnDefine => ({...p, isRelation: true}), undefined,
  (target: any, name: string, columnDefine: ColumnDefine) => {

  });


// // tslint:disable-next-line:variable-name
// export const Relation: ColumnDecorator = makePropDecorator(
//   'fedaco orm relation',
//   (p: RelationAnnotation = {}) => ({...p}), undefined,
//   (target: any, name: string, columnDefine) => {
//     additionalProcessing(target, name, columnDefine);
//   }
// );

export const ManyToManyRelation: ColumnDecorator = makePropDecorator(
  'fedaco orm many to many relation',
  (p: RelationAnnotation = {}) => ({...p}), undefined,
  (target: any, name: string, columnDefine) => {
    _additionalProcessing(target, name, columnDefine);
  }
);

export const OneToManyRelation: ColumnDecorator = makePropDecorator(
  'fedaco orm many to many relation',
  (p: RelationAnnotation = {}) => ({...p}), undefined,
  (target: any, name: string, columnDefine) => {
    _additionalProcessing(target, name, columnDefine);
  }
);


