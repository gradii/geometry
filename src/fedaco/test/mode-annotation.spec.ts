import { reflector } from '@gradii/annotation';
import { findLast } from 'ramda';
import { Column } from '../src/annotation/column';
import { Model } from '../src/fedaco/model';
import { BasicModel, } from './model/basic.model';
import { RelationModel } from './model/relation.model';

describe('model annotation', () => {
  it('instance model', () => {
    const m = new BasicModel();

    expect(m).toBeInstanceOf(Model);
  });

  it('test basic', () => {
    const basic = new BasicModel();

    const metadata = reflector.propMetadata(BasicModel);

    expect(metadata['name'].length).toBe(1);
    expect(metadata['name'][0].columnName).toBe('name');
  });

  it('test basic model 1', () => {
    const basic = new BasicModel();
    basic.name  = 'hello';

    const meta = reflector.propMetadata(BasicModel);
    const a    = findLast(it => Column.isTypeOf(it), meta['name']);
  });

  it('test relation annoation', () => {
    const relationModel = new RelationModel();

    const meta = reflector.propMetadata(RelationModel);

    expect(meta).toMatchSnapshot('meta');
  });

  it('test relation annotation should call getAttribute', () => {

    const relationModel = new RelationModel();

    const meta = reflector.propMetadata(RelationModel);

    const spy1 = jest.spyOn(relationModel, 'getAttribute').mockReturnValue('foo');
    const data = relationModel.columnFoo;

    expect(spy1).toBeCalled();
    expect(data).toBe('foo');

  });


});
