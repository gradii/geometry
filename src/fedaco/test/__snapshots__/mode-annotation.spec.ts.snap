// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`model annotation test relation annoation should call getAttribute: meta 1`] = `
Object {
  "columnFoo": Array [
    PropDecoratorFactory {
      "isRelation": true,
    },
  ],
}
`;

exports[`model annotation test relation annoation: meta 1`] = `
Object {
  "columnFoo": Array [
    PropDecoratorFactory {
      "isRelation": true,
    },
  ],
}
`;
